cpp/src/strings/convert/convert_durations.cu

```diff
@@ -0,0 +1,650 @@
+      case 'H': return count_digits(timeparts[DU_HOUR]); break;
+      case 'M': return count_digits(timeparts[DU_MINUTE]); break;
+      case 'S': return count_digits(timeparts[DU_SECOND]); break;
+      // 0 or 6 digits for pandas, include dot only if non-zero.
+      case 'u': return (timeparts[DU_SUBSECOND] == 0) ? 0 : 6 + 1; break;
+      // 0 or ns without trailing zeros
+      case 'f':
+        return (timeparts[DU_SUBSECOND] == 0) ? 0 : [units = type] {
+          if (units == type_id::DURATION_MILLISECONDS) return 3 + 1;  // +1 is for dot
+          if (units == type_id::DURATION_MICROSECONDS) return 6 + 1;  // +1 is for dot
+          if (units == type_id::DURATION_NANOSECONDS) return 9 + 1;   // +1 is for dot
+          return 0;
+        }() - count_trailing_zeros(timeparts[DU_SUBSECOND]);  // 3/6/9-trailing_zeros.
+        break;
+      default: return 2;
+    }
+  }
+
+  __device__ size_type operator()(size_type idx)
+  {
+    if (d_durations.is_null(idx)) return 0;
+    auto duration                   = d_durations.element<T>(idx);
+    int32_t timeparts[DU_ARRAYSIZE] = {0};  // days, hours, minutes, seconds, subseconds(9)
+    dissect_duration(duration.count(), timeparts, type);
+    size_type string_length{0};
+    for (auto i = 0; i < items_count; i++) {
+      format_item item = d_format_items[i];
+      if (item.item_type == format_char_type::literal)
+        string_length++;
+      else if (item.length != -1)
+        string_length += item.length;
+      else
+        string_length += format_length(item.value, timeparts);
+    }
+    return string_length;
+  }
+};
+
+template <typename T>
+struct duration_to_string_fn : public duration_to_string_size_fn<T> {
+  const int32_t* d_offsets;
+  char* d_chars;
+  using duration_to_string_size_fn<T>::d_durations;
+  using duration_to_string_size_fn<T>::d_format_items;
+  using duration_to_string_size_fn<T>::items_count;
+  using duration_to_string_size_fn<T>::type;
+
+  duration_to_string_fn(const column_device_view d_durations,
+                        const format_item* d_format_items,
+                        size_type items_count,
+                        type_id type,
+                        const int32_t* d_offsets,
+                        char* d_chars)
+    : duration_to_string_size_fn<T>{d_durations, d_format_items, items_count, type},
+      d_offsets(d_offsets),
+      d_chars(d_chars)
+  {
+  }
+
+  // utility to create (optionally) 0-padded integers (up to 10 chars) with negative sign.
+  // min_digits==-1 indicates no 0-padding.
+  __device__ char* int2str(char* str, int min_digits, int32_t value)
+  {
+    constexpr int MAX_DIGITS = 10;  // largest 32-bit integer is 10 digits
+    assert(min_digits <= MAX_DIGITS);
+    if (value == 0) {
+      do {
+        *str++ = '0';
+      } while (--min_digits > 0);
+      return str;
+    }
+    bool is_negative = (value < 0);
+
+    char digits[MAX_DIGITS] = {'0', '0', '0', '0', '0', '0', '0', '0', '0', '0'};
```

Comment:
Here's another array where accesses are going to go to global memory.

Reply:
This is a fixed sized, 10-byte, array on the stack. Why would it be placed in global memory?

Reply:
In CUDA, any local arrays are placed in local memory - which are stored in GPU DRAM. (could be cached too). Also, any local variables if they can't be stored in registers, spill to DRAM. (I think global memory is loosed used term for GPU DRAM).

Reply:
> any local arrays are placed in local memory

The only caveat is that if the access pattern to the thread local array is known at compile time, then the compiler will put it into registers instead of local memory ("local" memory is misleading since it is actually stored in frame buffer memory).

Reply:
>  if the access pattern to the thread local array is known at compile time

... AND the array is small enough.  I can't remember what the threshold size is, but it's not very big.

Reply:
If this becomes a problem, profile it and see if putting it in shared memory (10 bytes per thread, not actually shared) helps performance. Might need to pad it out to avoid shmem bank conflicts.

Reply:
This is called through the `thrust::transform` so using shared memory may be difficult since we are not actually launching the kernel directly.